filip/semantics/vocabulary/relation.py

```diff
@@ -0,0 +1,276 @@
+from typing import List, Set, Dict, TYPE_CHECKING
+
+from pydantic import BaseModel
+from enum import Enum
+
+
+from . import TargetStatement, StatementType
+from .source import DependencyStatement
+
+if TYPE_CHECKING:
+    from . import Vocabulary
+
+
+class RestrictionType(str, Enum):
+    """RestrictionTypes, as defined for OWL:
+        some: at least 1 value of that target
+        only: only value of that target
+        min: min n values of that target
+        max: max n values of that target
+        exactly: exactly n values of that target
+        value: predefined value
+    """
+    some = 'some'
+    only = 'only'
+    min = 'min'
+    max = 'max'
+    exactly = 'exactly'
+    value = 'value'
+
+
+class Relation(BaseModel):
+    """
+    A Relation is defined in the source for a class.
+    It has the form: RestrictionType property target_statement
+
+    It defines a set of allowed/required values which each instance of this
+    class can/should have under this property
+
+    A Relation is defined in a OWL:class, but all child classes of that class
+    inherit it
+    """
+
+    id: str
+    """Unique generated Relation ID, for internal use"""
+    restriction_type: RestrictionType = None
+    """Restriction type of this relation"""
+    restriction_cardinality: int = -1
+    """Only needed for min, max, equaly states the 'n'"""
+    property_iri: str = ""
+    """IRI of the property (data- or object-)"""
+    target_statement: TargetStatement = None
+    """Complex statement which classes/datatype_catalogue are allowed/required"""
+
+    def get_targets(self) -> List[List[str]]:
+        """Get all targets specified in the target statement in AND-OR Notation
+
+        Returns:
+            List[List[str]]: [[a,b],[c]] either a and b needs to present, or c
+        """
+        return self.target_statement.get_all_targets()
+
+    def to_string(self, vocabulary: 'Vocabulary') -> str:
+        """ Get a string representation of the relation
+
+        Args:
+            vocabulary (Vocabulary): Vocabulary of this project
+
+        Returns:
+            str
+        """
+
+        if self.restriction_cardinality == -1:
+            return "{} {}".format(self.restriction_type, self.target_statement.
+                                  to_string(vocabulary))
+        else:
+            return self.restriction_type + " " + \
+                   str(self.restriction_cardinality) + " " \
+                   + self.target_statement.to_string(vocabulary)
+
+    def is_restriction_fulfilled(self, number_of_fulfilling_values: int,
+                                 total_number_of_values: int) -> bool:
```

Comment:
ssome of the function coudl be validators right?

Reply:
I would say no. As this validation and pydantic validators work on tow different levels. 
A pydantic validator checks if the object is allowed to exist with given values, this validator operates one level higher on semantic reasoning.

Further the methode is not used in the object itself